mqtt/mqtt-broker/src/persist.rs

```diff
@@ -1,28 +1,385 @@
+                        Err(Error::from(ErrorKind::Persist(ErrorReason::FileRename)))
+                    });
+                    fs::rename(&temp_link_path, &link_path)
+                        .context(ErrorKind::Persist(ErrorReason::FileRename))?;
+
+                    // Prune old states
+                    fail_point!("filepersistor.store.readdir", |_| {
+                        Err(Error::from(ErrorKind::Persist(ErrorReason::ReadDir)))
+                    });
+
+                    let mut entries = fs::read_dir(&dir)
+                        .context(ErrorKind::Persist(ErrorReason::ReadDir))?
+                        .filter_map(Result::ok)
+                        .filter(|entry| entry.file_type().ok().map_or(false, |f| f.is_file()))
+                        .filter(|entry| {
+                            entry
+                                .file_name()
+                                .to_string_lossy()
+                                .starts_with(STATE_DEFAULT_STEM)
+                        })
+                        .collect::<Vec<fs::DirEntry>>();
+
+                    entries.sort_unstable_by(|a, b| {
+                        b.file_name()
+                            .partial_cmp(&a.file_name())
+                            .unwrap_or(cmp::Ordering::Equal)
+                    });
+
+                    for entry in entries.iter().skip(previous_count) {
+                        debug!(
+                            "pruning old state file {}...",
+                            entry.file_name().to_string_lossy()
+                        );
+
+                        fail_point!("filepersistor.store.entry_unlink", |_| {
+                            Err(Error::from(ErrorKind::Persist(ErrorReason::FileUnlink)))
+                        });
+                        fs::remove_file(&entry.path())
+                            .context(ErrorKind::Persist(ErrorReason::FileUnlink))?;
+                        debug!("{} pruned.", entry.file_name().to_string_lossy());
+                    }
+                }
+                Err(e) => {
+                    fail_point!("filepersistor.store.new_file_unlink", |_| {
+                        Err(Error::from(ErrorKind::Persist(ErrorReason::FileUnlink)))
+                    });
+                    fs::remove_file(path).context(ErrorKind::Persist(ErrorReason::FileUnlink))?;
+                    return Err(e);
+                }
+            }
+            info!(message="persisted state.", file=%path.display());
+            Ok(())
+        })
+        .await;
+
+        fail_point!("filepersistor.store.spawn_blocking", |_| {
+            Err(Error::from(ErrorKind::TaskJoin))
+        });
+        res.context(ErrorKind::TaskJoin)?
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ErrorReason {
+    FileOpen,
+    FileRename,
+    FileUnlink,
+    CreateDir,
+    ReadDir,
+    Symlink,
+    SymlinkUnlink,
+    Serialize,
+    Deserialize,
+}
+
+impl fmt::Display for ErrorReason {
```

Comment:
In my "yet another error crate" I tried to get away with [derive_more](https://docs.rs/derive_more/0.99.3/derive_more/). Almost succedded :)